playwright-tests/tests/system-updates/update-widgets.spec.js

```diff
@@ -0,0 +1,255 @@
+import { expect } from "@playwright/test";
+import { cacheCDN, test } from "../../util/test.js";
+import {
+  DEFAULT_WIDGET_REFERENCE_ACCOUNT_ID,
+  SandboxRPC,
+} from "../../util/sandboxrpc.js";
+import { createDAOargs } from "../../util/sputnikdao.js";
+import nearApi from "near-api-js";
+import { readFile } from "fs/promises";
+
+test("should update bootstrap widget and upgrade instance with it", async ({
+  page,
+}) => {
+  test.setTimeout(120_000);
+
+  await cacheCDN(page);
+
+  const sandbox = new SandboxRPC();
+  await sandbox.init();
+
+  const widget_reference_account_id = DEFAULT_WIDGET_REFERENCE_ACCOUNT_ID;
+  await sandbox.setupDefaultWidgetReferenceAccount();
+
+  const instanceName = "widgetupdater";
+
+  const instanceAccountId = `${instanceName}.near`;
+
+  const createInstanceResult = await sandbox.account.functionCall({
+    contractId: "treasury-factory.near",
+    methodName: "create_instance",
+    args: {
+      sputnik_dao_factory_account_id: "sputnik-dao.near",
+      social_db_account_id: "social.near",
+      widget_reference_account_id: widget_reference_account_id,
+      name: instanceName,
+      create_dao_args: Buffer.from(
+        JSON.stringify(
+          createDAOargs({
+            instanceName: instanceName,
+            adminAccountId: sandbox.account.accountId,
+          })
+        )
+      ).toString("base64"),
+    },
+    gas: 300000000000000,
+    attachedDeposit: nearApi.utils.format.parseNearAmount("9"),
+  });
+
+  expect(
+    createInstanceResult.receipts_outcome.filter(
+      (receipt_outcome) => receipt_outcome.outcome.status.Failure
+    ).length
+  ).toBe(0);
+
+  // The initial update that is already applied, so should automatically be dismissed when visiting the updates page
+  await sandbox.modifyWidget(
+    "widgets.treasury-factory.near/widget/pages.settings.system-updates.UpdateRegistry",
+    `
+    return [
+      {
+        id: 1,
+        createdDate: "2025-03-25",
+        version: "n/a",
+        type: "Widgets",
+        summary: "Added lockup",
+        details: "",
+        votingRequired: false,
+      }
+  ];
+  `
+  );
+
+  await sandbox.redirectWeb4(instanceAccountId, page);
+
+  await page.goto(`https://${instanceName}.near.page`);
+
+  // Normal users should not see the update banner
+  await page.waitForTimeout(500);
+  await expect(
+    await page.getByRole("link", { name: "Review" })
+  ).not.toBeVisible();
+
+  const keyPair = await sandbox.keyStore.getKey("sandbox", sandbox.account_id);
+  await page.evaluate(
+    ({ accountId, publicKey, privateKey }) => {
```

Comment:
I remember you created a reusable function or something in earlier PR to set all the local storage? Am I missing something?

Reply:
The only alternative I can recall is the storage state ( from Playwright ), but that does not support dynamic values. Here we are creating different accounts each time, and the keys needs to be stored in the browsers localstorage, I cannot find that we have a reusable function for this right now.